pkg/provisioner/ironic/validatemanagementaccess_test.go

```diff
@@ -0,0 +1,97 @@
+
+	"github.com/metal3-io/baremetal-operator/pkg/bmc"
+	"github.com/metal3-io/baremetal-operator/pkg/provisioner/ironic/clients"
+	"github.com/metal3-io/baremetal-operator/pkg/provisioner/ironic/testserver"
+)
+
+func TestValidateManagementAccessNoMAC(t *testing.T) {
+	// Create a host without a bootMACAddress and with a BMC that
+	// requires one.
+	host := makeHost()
+	host.Spec.BMC.Address = "test-needs-mac://"
+	host.Spec.BootMACAddress = ""
+	host.Status.Provisioning.ID = "" // so we don't lookup by uuid
+
+	ironic := testserver.NewIronic(t).Ready().NoNode(host.Name)
+	ironic.Start()
+	defer ironic.Stop()
+
+	auth := clients.AuthConfig{Type: clients.NoAuth}
+	prov, err := newProvisionerWithSettings(host, bmc.Credentials{}, nil,
+		ironic.Endpoint(), auth, testserver.NewInspector(t).Endpoint(), auth,
+	)
+	if err != nil {
+		t.Fatalf("could not create provisioner: %s", err)
+	}
+
+	result, err := prov.ValidateManagementAccess(false)
+	if err != nil {
+		t.Fatalf("error from ValidateManagementAccess: %s", err)
+	}
+	assert.Contains(t, result.ErrorMessage, "requires a BootMACAddress")
+}
+
+func TestValidateManagementAccessMACOptional(t *testing.T) {
+	// Create a host without a bootMACAddress and with a BMC that
+	// does not require one.
+	host := makeHost()
+	host.Spec.BootMACAddress = ""
+
+	// Set up ironic server to return the node
+	ironic := testserver.NewIronic(t).Ready().
+		WithNode(nodes.Node{
+			Name: host.Name,
+			UUID: host.Status.Provisioning.ID,
+		})
+	ironic.Start()
+	defer ironic.Stop()
+
+	auth := clients.AuthConfig{Type: clients.NoAuth}
+	prov, err := newProvisionerWithSettings(host, bmc.Credentials{}, nil,
+		ironic.Endpoint(), auth, testserver.NewInspector(t).Endpoint(), auth,
+	)
+	if err != nil {
+		t.Fatalf("could not create provisioner: %s", err)
+	}
+
+	result, err := prov.ValidateManagementAccess(false)
+	if err != nil {
+		t.Fatalf("error from ValidateManagementAccess: %s", err)
+	}
+	assert.Equal(t, "", result.ErrorMessage)
+}
+
+func TestValidateManagementAccessCreateNode(t *testing.T) {
+	// Create a host without a bootMACAddress and with a BMC that
+	// does not require one.
+	host := makeHost()
+	host.Spec.BootMACAddress = ""
+	host.Status.Provisioning.ID = "" // so we don't lookup by uuid
+
+	ironic := testserver.NewIronic(t).Ready().CreateNodes().NoNode(host.Name)
+	ironic.Start()
+	defer ironic.Stop()
+
+	auth := clients.AuthConfig{Type: clients.NoAuth}
+	prov, err := newProvisionerWithSettings(host, bmc.Credentials{}, nullEventPublisher,
+		ironic.Endpoint(), auth, testserver.NewInspector(t).Endpoint(), auth,
+	)
+	if err != nil {
+		t.Fatalf("could not create provisioner: %s", err)
+	}
+
+	result, err := prov.ValidateManagementAccess(false)
+	if err != nil {
+		t.Fatalf("error from ValidateManagementAccess: %s", err)
+	}
+	assert.Equal(t, "", result.ErrorMessage)
+	assert.NotEqual(t, "", host.Status.Provisioning.ID)
+	assert.Equal(t, ironic.CreatedNodes[0].UUID, host.Status.Provisioning.ID)
+}
```

Comment:
(Not blocking) I noticed that the three test cases have been split into three different test methods (`TestValidateManagementAccessNoMAC`, `TestValidateManagementAccessMACOptional` and `TestValidateManagementAccessCreateNode`) sharing a pretty similar setup code.
Maybe it's not worth in this case, but to transform the three methods in a list of cases in a single test method it could be useful to "upgrade" the `makeHost()` factory method to a proper `host builder` object with fluent interface, so that also the host could become a parameter of the cases list (and of course also the expectations code will need to be adjusted).

Reply:
Given that the `ValidateManagementAccess` is a pretty big method to be covered, the existing node and/or credentials change case will be added within this PR?

Reply:
The setup portions of the functions are similar, but the checks were different enough that it seemed like it would be clearer to just write separate functions this time, instead of having more complicated logic to decide which tests to apply in each case.

In terms of test coverage, I think this PR is big enough and I would like to merge it before finishing the tests for ValidateManagementAccess.

Reply:
Yes, makes sense